From 59dec4d0cb9eacf248c10f463ba48fd204d9c6ae Mon Sep 17 00:00:00 2001
From: Jacky Bai <ping.bai@nxp.com>
Date: Wed, 11 Sep 2019 10:12:47 +0800
Subject: [PATCH] plat: imx8m: Fix the derate enable conflict issue

The derate enable setting should use save/restore
to make sure the enable control is consistent with
the initial setting.

Signed-off-by: Jacky Bai <ping.bai@nxp.com>
---
 plat/imx/common/imx8m/lpddr4_dvfs.c      | 10 +++++++---
 plat/imx/common/imx8m/lpddr4_retention.c |  2 --
 2 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/plat/imx/common/imx8m/lpddr4_dvfs.c b/plat/imx/common/imx8m/lpddr4_dvfs.c
index b3ea27bb7..6fee4496a 100644
--- a/plat/imx/common/imx8m/lpddr4_dvfs.c
+++ b/plat/imx/common/imx8m/lpddr4_dvfs.c
@@ -20,6 +20,7 @@ void lpddr4_swffc(struct dram_info *info, unsigned int init_fsp,
 	unsigned int mr, emr, emr2, emr3;
 	unsigned int mr11, mr12, mr22, mr14;
 	unsigned int tmp;
+	unsigned int derate_backup[3];
 
 	/* 1. program targetd UMCTL2_REGS_FREQ1/2/3,already done, skip it. */
 
@@ -109,14 +110,17 @@ void lpddr4_swffc(struct dram_info *info, unsigned int init_fsp,
 
 	/* 10. Disable automatic derating: derate_enable */
 	tmp= mmio_read_32(DDRC_DERATEEN(0));
+	derate_backup[0] = tmp;
 	tmp &= ~0x1;
 	mmio_write_32(DDRC_DERATEEN(0), tmp);
 
 	tmp= mmio_read_32(DDRC_FREQ1_DERATEEN(0));
+	derate_backup[1] = tmp;
 	tmp &= ~0x1;
 	mmio_write_32(DDRC_FREQ1_DERATEEN(0), tmp);
 
 	tmp= mmio_read_32(DDRC_FREQ2_DERATEEN(0));
+	derate_backup[2] = tmp;
 	tmp &= ~0x1;
 	mmio_write_32(DDRC_FREQ2_DERATEEN(0), tmp);
 
@@ -347,9 +351,9 @@ void lpddr4_swffc(struct dram_info *info, unsigned int init_fsp,
 	}
 
 	/* 40. re-emable automatic derating: derate_enable */
-	tmp= mmio_read_32(DDRC_DERATEEN(0));
-	tmp &= 0xFFFFFFFE;
-	mmio_write_32(DDRC_DERATEEN(0), tmp);
+	mmio_write_32(DDRC_DERATEEN(0), derate_backup[0]);
+	mmio_write_32(DDRC_FREQ1_DERATEEN(0), derate_backup[1]);
+	mmio_write_32(DDRC_FREQ2_DERATEEN(0), derate_backup[2]);
 
 	/* 41. write 1 to PCTRL.port_en */
 	mmio_write_32(DDRC_PCTRL_0(0), 0x1);
diff --git a/plat/imx/common/imx8m/lpddr4_retention.c b/plat/imx/common/imx8m/lpddr4_retention.c
index 8cc44c891..1f77ccb9f 100644
--- a/plat/imx/common/imx8m/lpddr4_retention.c
+++ b/plat/imx/common/imx8m/lpddr4_retention.c
@@ -250,8 +250,6 @@ void lpddr4_exit_retention(void)
 		INFO("wait STAT to normal state\n");
 	}
  
-	mmio_write_32(DDRC_DERATEEN(0), 0x00000302);
-
 	mmio_write_32(DDRC_PCTRL_0(0), 0x00000001); 
 	 /* dis_auto-refresh is set to 0 */
 	mmio_write_32(DDRC_RFSHCTL3(0), 0x00000000);
-- 
2.25.0

